src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,323 @@
+
+    // Solve fully general case
+    inline CELER_FUNCTION Intersections operator()(Real3 const& pos,
+                                                   Real3 const& dir) const;
+
+    //// ACCESSORS ////
+
+    //! Get involute parameters
+    CELER_FUNCTION real_type r_b() const { return r_b_; }
+    CELER_FUNCTION real_type a() const { return a_; }
+    CELER_FUNCTION real_type sign() const { return mirror_; }
+
+    //! Get bounds of the involute
+    CELER_FUNCTION real_type tmin() const { return tmin_; }
+    CELER_FUNCTION real_type tmax() const { return tmax_; }
+
+  private:
+    //// DATA ////
+    // Involute parameters
+    real_type r_b_;
+    real_type a_;
+    bool mirror_;
+
+    // Bounds
+    real_type tmin_;
+    real_type tmax_;
+};
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from involute parameters.
+ */
+CELER_FUNCTION InvoluteSolver::InvoluteSolver(
+    real_type r_b, real_type a, real_type sign, real_type tmin, real_type tmax)
+    : r_b_(r_b), a_(a), mirror_(sign < 0), tmin_(tmin), tmax_(tmax)
+{
+    CELER_EXPECT(r_b > 0);
+    CELER_EXPECT(a > 0);
+    CELER_EXPECT(std::fabs(tmax) < 2 * pi + std::fabs(tmin));
+
+    if (mirror_)
+    {
+        a_ = -a + pi;
+    }
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Find all roots for involute surfaces that are within the bounds and result
+ * in positive distances. Performed by doing a Regular Falsi Iteration on the
+ * root function, \f[
+ * f(t) = r_b * [v{cos(a+t) + tsin(a+t)} + u{sin(a+t) - tcos(a+t)}] + xv - yu
+ * \f]
+ * where the Regular Falsi Iteration is given by: \f[
+ * tc = [ta*f(tb) - tb*f(ta)] / [f(tb) - f(ta)]
+ * \f]
+ * where \em tc replaces the bound with the same sign (e.g. \em ta and \em tb
+ * ). The initial bounds can be determined by the set of: /f[ {0, beta - a,
+ * beta - a - pi, beta - a + pi, beta - a - 2pi, beta - a + 2pi ...} /f] Where
+ * \em beta is: \f[ beta = arctan(-v/u) \f]
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(Real3 const& pos,
+                           Real3 const& dir) const -> Intersections
+{
+    // Flatten pos and dir in xyz and uv respectively
+    real_type x = pos[0];
+    real_type const y = pos[1];
+    real_type const z = pos[2];
+
+    real_type u = dir[0];
+    real_type const v = dir[1];
+    real_type const w = dir[2];
+
+    if (mirror_)
+    {
+        x = -x;
+        u = -u;
+    }
+
+    // Lambda used for calculating the roots using Regular Falsi Iteration
+    auto root = [](real_type t,
+                   real_type x,
+                   real_type y,
+                   real_type u,
+                   real_type v,
+                   real_type r_b_,
+                   real_type a_) {
```

Comment:
remove the underscores from the function parameters.  Better yet, just capture `this` and then you can use those guys without passing them in.
`[this](...)`